pkg/utils/resource/update/update.go

```diff
@@ -0,0 +1,72 @@
+package update
+
+import (
+	"context"
+	"reflect"
+
+	cco "github.com/openshift/cloud-credential-operator/pkg/apis/cloudcredential/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/openshift/aws-load-balancer-operator/pkg/utils"
+)
+
+// CredentialsRequest updates the given CredentialsRequest to the desired state if needed.
+// Returns true if the CredentialsRequest was updated, false otherwise.
+func CredentialsRequest(ctx context.Context, client client.Client, current *cco.CredentialsRequest, desired *cco.CredentialsRequest) (bool, error) {
+	if changed, err := IsCredentialsRequestChanged(current, desired); err != nil {
+		return false, err
+	} else if !changed {
+		return false, nil
+	}
+
+	updated := current.DeepCopy()
+	updated.Name = desired.Name
+	updated.Namespace = desired.Namespace
+	updated.Spec = desired.Spec
+	if err := client.Update(ctx, updated); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// IsCredentialsRequestChanged compares the current and desired states of a CredentialsRequest.
+// Returns true of current start is different from the desired, false otherwise.
+func IsCredentialsRequestChanged(current, desired *cco.CredentialsRequest) (bool, error) {
+	if current.Name != desired.Name {
+		return true, nil
+	}
+
+	if current.Namespace != desired.Namespace {
+		return true, nil
+	}
+
+	if !reflect.DeepEqual(current.Spec.SecretRef, desired.Spec.SecretRef) {
+		return true, nil
+	}
+
+	if !utils.EqualStrings(current.Spec.ServiceAccountNames, desired.Spec.ServiceAccountNames) {
+		return true, nil
+	}
+
+	if current.Spec.CloudTokenPath != desired.Spec.CloudTokenPath {
+		return true, nil
+	}
+
+	currentAwsSpec := cco.AWSProviderSpec{}
+	err := cco.Codec.DecodeProviderSpec(current.Spec.ProviderSpec, &currentAwsSpec)
```

Comment:
Just curious, do you need to decode provider spec? What happens when you compare them as Raw?

Reply:
I thought of the readability, the decoded diff is more human readable. However I tried raw too, it has a small glitch which I didn't dig for the sake of time as I was fine with the decoded approach - there is a new line which gets inserted somewhere in the CCO code.